cab-token-generator/javatests/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactoryTest.java

```diff
@@ -0,0 +1,573 @@
+        .setClock(createMockClock(refreshType, sourceCredentials))
+        .build();
+  }
+
+  private Clock createMockClock(RefreshType refreshType, GoogleCredentials sourceCredentials) {
+    Clock mockClock = mock(Clock.class);
+    long currentTimeInMillis = Clock.SYSTEM.currentTimeMillis();
+    long mockedTimeInMillis;
+    final long refreshMarginInMillis =
+        ClientSideCredentialAccessBoundaryFactory.DEFAULT_REFRESH_MARGIN.toMillis();
+    final long minimumTokenLifetimeMillis =
+        ClientSideCredentialAccessBoundaryFactory.DEFAULT_MINIMUM_TOKEN_LIFETIME.toMillis();
+
+    // If the source credential doesn't have an access token, set the expiration time to 1 hour from
+    // the current time.
+    long expirationTimeInMillis =
+        sourceCredentials.getAccessToken() != null
+            ? sourceCredentials.getAccessToken().getExpirationTime().getTime()
+            : currentTimeInMillis + 3600000;
+
+    switch (refreshType) {
+      case NONE:
+        // Set mocked time so that the token is fresh and no refresh is needed (before the refresh
+        // margin).
+        mockedTimeInMillis = expirationTimeInMillis - refreshMarginInMillis - 60000;
+        break;
+      case ASYNC:
+        // Set mocked time so that the token is nearing expiry and an async refresh is triggered
+        // (within the refresh margin).
+        mockedTimeInMillis = expirationTimeInMillis - refreshMarginInMillis + 60000;
+        break;
+      case BLOCKING:
+        // Set mocked time so that the token requires immediate refresh (just after the minimum
+        // token lifetime).
+        mockedTimeInMillis = expirationTimeInMillis - minimumTokenLifetimeMillis + 60000;
+        break;
+      default:
+        throw new IllegalArgumentException("Unexpected RefreshType: " + refreshType);
+    }
+
+    when(mockClock.currentTimeMillis()).thenReturn(mockedTimeInMillis);
+    return mockClock;
+  }
+
+  private static void triggerConcurrentRefresh(
+      ClientSideCredentialAccessBoundaryFactory factory, int numThreads)
+      throws InterruptedException {
+    Thread[] threads = new Thread[numThreads];
+    CountDownLatch latch = new CountDownLatch(numThreads);
+
+    // Create and start the threads
+    for (int i = 0; i < numThreads; i++) {
+      threads[i] =
+          new Thread(
+              () -> {
+                try {
+                  latch.countDown(); // Signal thread is ready
+                  latch.await(); // Wait for all threads to be ready
+                  factory.refreshCredentialsIfRequired();
+                } catch (InterruptedException e) {
+                  Thread.currentThread().interrupt();
+                } catch (IOException e) {
+                  throw new RuntimeException(e);
+                }
+              });
+      threads[i].start();
+    }
+    for (Thread thread : threads) {
+      thread.join(); // Wait for each thread to complete
+    }
```

Comment:
Can this join() have a timeout? Say like 5 seconds for the tests so they don't run forever in the CI?